warehouse/oidc/views.py

```diff
@@ -0,0 +1,110 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from pyramid.view import view_config
+
+from warehouse.admin.flags import AdminFlagValue
+from warehouse.events.tags import EventTag
+from warehouse.macaroons import caveats
+from warehouse.macaroons.interfaces import IMacaroonService
+from warehouse.oidc.interfaces import IOIDCProviderService
+
+
+@view_config(
+    route_name="oidc.mint_token",
+    require_methods=["POST"],
+    renderer="json",
+    require_csrf=False,
+    has_translations=False,
+)
+def mint_token_from_oidc(request):
+    def _invalid(errors):
+        request.response.status = 422
+        return {"message": "Token request failed", "errors": errors}
+
+    oidc_enabled = request.registry.settings[
+        "warehouse.oidc.enabled"
+    ] and not request.flags.enabled(AdminFlagValue.DISALLOW_OIDC)
+    if not oidc_enabled:
+        return _invalid(
+            errors=[
+                {
+                    "code": "not-enabled",
+                    "description": "OIDC functionality not enabled",
+                }
+            ]
+        )
+
+    try:
+        body = request.json_body
+    except ValueError:
+        return _invalid(
+            errors=[{"code": "invalid-json", "description": "missing JSON body"}]
+        )
+
+    # `json_body` can return any valid top-level JSON type, so we have
+    # to make sure we're actually receiving a dictionary.
+    if not isinstance(body, dict):
+        return _invalid(
+            errors=[
+                {
+                    "code": "invalid-payload",
+                    "description": "payload is not a JSON dictionary",
+                }
+            ]
+        )
+
+    unverified_jwt = body.get("token")
+    if not unverified_jwt or not isinstance(unverified_jwt, str):
+        return _invalid(
+            errors=[{"code": "invalid-token", "description": "missing or empty token"}]
```

Comment:
Can we make this error more clear? Two things could be going wrong here: the 'token' field is missing entirely, or the value in the 'token' field is invalid.

Reply:
Yep! Doing now.

Reply:
e2df77ee